js/src/forum/components/SsoGrid.js

```diff
@@ -0,0 +1,147 @@
+    this.providers = this.ssoProviders().toArray();
+
+    /**
+     * A map of Sso type-method combinations to the checkbox instances
+     * that represent them.
+     *
+     * @type {Object}
+     */
+    this.inputs = {};
+
+    // For each of the Sso type-method combinations, create and store a
+    // new checkbox component instance, which we will render in the view.
+    this.providers.forEach(provider => {
+      this.inputs[provider.name] = Checkbox.component({
+        state: this.providerLinked(provider.name),
+        disabled: this.disableUnlinking(provider.name),
+        onchange: () => this.toggle(provider.name),
+      });
+    });
+  }
+
+  view() {
+    return (
+      <table className="SsoGrid">
+        <thead>
+          <tr>
+            <td />
+            <th className="SsoGrid-groupToggle">
+              {app.translator.trans('core.forum.settings.sso.linked_header')}
+            </th>
+          </tr>
+        </thead>
+
+        <tbody>
+          {this.providers.map(provider => (
+            <tr>
+              <td className="SsoGrid-groupToggle">
+                {icon(provider.icon)} {provider.label}
+              </td>
+              <td className="SsoGrid-checkbox">
+                {this.inputs[provider.name]}
+              </td>
+            </tr>
+          ))}
+        </tbody>
+      </table>
+    );
+  }
+
+  providerLinked(provider) {
+    return app.session.user.ssoProviders().includes(provider);
+  }
+
+  disableUnlinking(provider) {
+    return this.providerLinked(provider) && !app.forum.attribute('enablePasswordAuth') && app.session.user.ssoProviders().length === 1;
+  }
+
+  /**
+   * Toggle the linked/unlinked state of the given sso provider.
+   *
+   * @param {Array} keys
+   */
+  toggle(provider) {
+    const control = this.inputs[provider];
+    control.loading = true
+
+    if (this.providerLinked(provider)) {
+      m.redraw();
+
+      app.request({
+        method: 'DELETE',
+        url: app.forum.attribute('apiUrl') + '/auth/' + provider
+      })
+        .then(() => {
+          control.props.state = false;
+          control.loading = false;
+          app.session.user.data.attributes.ssoProviders = app.session.user.ssoProviders().filter(item => item != provider);
+          for (const input in this.inputs) {
+            this.inputs[input].props.disabled = this.disableUnlinking(input);
+          }
+          m.redraw();
+        })
+        .catch(() => {
+          control.loading = false;
+          m.redraw();
+        });
+    } else {
+      const width = 580;
+      const height = 400;
+      const $window = $(window);
+
+      window.open(app.forum.attribute('baseUrl') + '/auth/' + provider, 'logInPopup',
+        `width=${width},` +
+        `height=${height},` +
+        `top=${$window.height() / 2 - height / 2},` +
+        `left=${$window.width() / 2 - width / 2},` +
+        'status=no,scrollbars=yes,resizable=no');
+      control.loading = false;
+      m.redraw();
```

Comment:
The two `m.redraw()` in `if` and `else` could be moved outside of the condition block and replaced by a single call at the end of the `toggle()` method.

Maybe the redraw can be completely removed because Mithril is supposed to already redraw after an `onchange` event like the one that's used to call `toggle`.

Reply:
The ones in the if statement need to be there since it's an asynchronous operation. For the redraw in else(), we could remove it, but then if the user closes the window it'll just keep on loading. I'm actually not 100% sure whether it should spin if a new provider is being linked.

Reply:
Oops I misspoke. The other redraw I saw was line 78, which is before the condition. If you move the redraw from line 78 to the end of the `toggle()` method, it will take care of the one you removed as well.